components/sup/src/manager/user_config_watcher.rs

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::thread::Builder as ThreadBuilder;
+
+use super::file_watcher::{Callbacks, default_file_watcher};
+
+use hcore::fs::USER_CONFIG_FILE;
+use hcore::service::ServiceGroup;
+use manager::service::Service;
+
+static LOGKEY: &'static str = "UCM";
+
+// This trait exists to ease the testing of functions that receive a Service.  Creating Services
+// requires a lot of ceremony, so we work around this with this trait.
+pub trait Serviceable {
+    fn name(&self) -> &String;
```

Comment:
Returning references makes the trait less flexible, because you need an existing, non-temporary thing in memory to return its address (like a member of a struct that implements this trait). In other words - you can't compute the return value. Not sure if it is really important, but in general I would just return plain values. Of course the downside of returning values is that you need to clone the stuff, but OTOH this won't likely be a bottleneck ever.

Reply:
My rationale here is that the name is always a reference to its containing struct.

Isn't it enough for users of this trait to be able to call `name.clone()` if they want to own the `String`?

Can you show an example of where the current implementation creates issues? As I'm not sure I understand your comment.

Reply:
There are no problems with current implementation. It was just a general note about returning references in trait functions. So in this case it's likely ok, but in general I would avoid it.

Reply:
OK, let me know if you feel strongly about changing this, otherwise I'm gonna leave it as is.

Reply:
Nah, leave it as is, please. If somebody else points out this thing, then you can change it.